Core/clim-basic/input.lisp

```diff
@@ -409,22 +408,29 @@ use condition-variables nor locks."))
   (check-schedule queue)
   (not (null (event-queue-head queue))))
 
-(defmethod event-queue-listen-or-wait ((queue concurrent-event-queue) &key timeout)
+(defmethod event-queue-listen-or-wait ((queue concurrent-event-queue)
+                                       &key timeout wait-function)
   (do-port-force-output queue)
   ;; We need to LOOP because of possible spurious wakeup (sbcl/bt quirk).
   (loop
+     with pred = (or wait-function #'(lambda () (event-queue-head queue)))
      with lock = (event-queue-lock queue)
      with cv = (event-queue-processes queue)
      with timeout-time = (and timeout (+ timeout (now)))
      as decay = (compute-decay timeout-time (event-schedule-time queue))
      do
-       (when decay (maxf decay 0))
+     ;; We CLAMPF decay when wait-function is present to ensure that
+     ;; we don't get stuck until next event arrives (or the timeout
+     ;; happens). It is busy wait with a lousy grain. -- jd 2019-06-06
+       (when decay (if wait-function
+                       (clampf decay 0 0.01)
```

Comment:
no, we do not want to busy-wait (0.01 is a frequent check anyway), putting 0 would make cpu spin at 100%.

Reply:
Does this actually happen? If the decay is 0, then we've already passed the time where the waiting should end, and as such it wouldn't be called again.

Reply:
we loop here because if wait-function is specified it doesn't matter how many events are processed. if the decay is 0, then condition-wait returns immedietely (if it were NIL it would return on input), and if predicate returns NIL and we are still before the timeout, we loop over again with decay=0. so yes, it does.